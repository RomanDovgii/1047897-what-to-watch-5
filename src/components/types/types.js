import PropTypes from "prop-types";

const headerSettingsType = {
  isUserPage: PropTypes.bool,
  isMyList: PropTypes.bool,
  isSignIn: PropTypes.bool,
  isNavigation: PropTypes.bool,
  isUserBlock: PropTypes.bool
};

export const isRenderedType = {
  isRendered: PropTypes.bool.isRequired
};

const optionalFunctionType = PropTypes.func;
const functionType = PropTypes.func.isRequired;
const requiredString = PropTypes.string.isRequired;

export const movieObjectType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: requiredString,
  posterImage: requiredString,
  backgroundImage: requiredString,
  backgroundColor: requiredString,
  videoLink: requiredString,
  previewVideoLink: requiredString,
  description: requiredString,
  rating: PropTypes.number.isRequired,
  scoresCount: PropTypes.number.isRequired,
  director: requiredString,
  starring: PropTypes.arrayOf(requiredString).isRequired,
  runTime: PropTypes.number.isRequired,
  genre: requiredString,
  released: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired

}).isRequired;

export const commentType = PropTypes.shape({
  text: PropTypes.string.isRequired,
  rating: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  date: PropTypes.instanceOf(Date).isRequired
}).isRequired;

const commentsType = PropTypes.objectOf(commentType).isRequired;

export const movieType = {
  movie: movieObjectType
};

export const onExitButtonClickType = {
  onExitButtonClick: functionType
};

export const onUserIconClickType = {
  onUserIconClick: optionalFunctionType
};

const onPlayButtonClickType = {
  onPlayButtonClick: optionalFunctionType
};

export const headingType = {
  heading: PropTypes.string.isRequired
};

export const moviesType = {
  movies: PropTypes.arrayOf(movieObjectType).isRequired
};

export const listItemType = PropTypes.string.isRequired;
const optionalListItemType = PropTypes.string;

export const optionalGenresType = {
  genres: PropTypes.arrayOf(optionalListItemType)
};

export const optionalGenreType = {
  genre: optionalListItemType
};

const genreType = {
  genre: listItemType
};

const genresType = {
  genres: PropTypes.arrayOf(listItemType).isRequired
};

const selectedGenreType = {
  selectedGenre: PropTypes.string.isRequired
};

const onGenreClick = {
  onGenreClick: functionType
};

const isSelected = {
  isSelected: PropTypes.bool.isRequired
};

export const videoPlayerType = {
  videoUrl: PropTypes.string.isRequired,
  imageName: PropTypes.string.isRequired,
  callSource: PropTypes.string.isRequired
};

export const screenTabType = {
  screen: PropTypes.string.isRequired
};

const moviesCountType = {
  moviesCount: PropTypes.number.isRequired
};

const onMoreButtonClickType = {
  onMoreButtonClick: functionType
};

const onWTWLogoClickType = {
  onWTWLogoClick: functionType
};

const resetStateType = {
  resetState: functionType
};

export const childrenType = {
  children: PropTypes.node.isRequired
};

const renderPlayerType = {
  renderPlayer: functionType
};

const callSourceType = {
  callSource: PropTypes.string.isRequired
};

const imageNameType = {
  imageName: PropTypes.string.isRequired
};

const isPlayingType = {
  isPlaying: PropTypes.bool.isRequired
};

const videoUrlType = {
  videoUrl: PropTypes.string.isRequired
};

const onChangeType = {
  onChange: functionType
};

const onSubmitType = {
  onSubmit: functionType
};

const ratingType = {
  rating: PropTypes.string.isRequired
};

const reviewType = {
  review: PropTypes.string.isRequired
};

const onFullScreenButtonClickType = {
  onFullScreenButtonClick: functionType
};

const isLoadingType = {
  isLoading: PropTypes.bool.isRequired
};

const durationType = {
  duration: PropTypes.number.isRequired
};

const currentTimeType = {
  currentTime: PropTypes.number.isRequired
};

export const onBigPlayButtonClickType = {
  onBigPlayButtonClick: functionType
};

const selectedMovie = {
  selectedMovie: movieObjectType
};

const authoriationStatusType = {
  authorizationStatus: PropTypes.bool.isRequired
};

const exactType = {
  exact: PropTypes.bool.isRequired
};

const pathType = {
  path: PropTypes.string.isRequired
};

const renderType = {
  render: functionType
};

export const smallMovieCardType = Object.assign({}, movieObjectType);
export const promotedMovieMoviesCommentsType = Object.assign({}, movieObjectType, moviesType, commentsType);

export const mainPageType = Object.assign({}, onUserIconClickType, onWTWLogoClickType, onPlayButtonClickType);
export const moviePageType = Object.assign({}, onUserIconClickType, onWTWLogoClickType, onPlayButtonClickType, selectedMovie, moviesType, commentsType);
export const movieCardTopType = Object.assign({}, onUserIconClickType, onWTWLogoClickType, onPlayButtonClickType, movieType, commentsType);
export const headerType = Object.assign({}, onUserIconClickType, headerSettingsType);
export const userBlockType = Object.assign({}, onUserIconClickType, isRenderedType);
export const catalogType = Object.assign({}, headingType, moviesType, optionalGenresType);
export const moreButtonType = Object.assign({}, headingType, moviesCountType, onMoreButtonClickType);
export const myListPageType = Object.assign({}, onUserIconClickType, onWTWLogoClickType);
export const movieCardType = Object.assign({}, onUserIconClickType, onWTWLogoClickType, onPlayButtonClickType, selectedMovie, commentsType);
export const tabsType = Object.assign({}, screenTabType, movieObjectType, commentsType);
export const tabType = Object.assign({}, movieObjectType);
export const tabWithCommentsType = Object.assign({}, movieObjectType, commentsType);
export const genresListType = Object.assign({}, genresType, selectedGenreType, onGenreClick);
export const genreListItemType = Object.assign({}, genreType, onGenreClick, isSelected);
export const logoType = Object.assign({}, onWTWLogoClickType, resetStateType);
export const withActiveCardType = Object.assign({}, movieType, renderPlayerType);
export const withActivePlayerType = Object.assign({}, callSourceType, imageNameType, isPlayingType, videoUrlType);
export const withActiveTabsType = Object.assign({}, commentsType, onPlayButtonClickType, onUserIconClickType, onWTWLogoClickType, selectedMovie);
export const addReviewType = Object.assign({}, ratingType, reviewType, onChangeType, onSubmitType);
export const playerControlsType = Object.assign({}, onPlayButtonClickType, onFullScreenButtonClickType, isPlayingType, isLoadingType, durationType, currentTimeType);
export const playerPageType = Object.assign({}, movieType, isPlayingType, onExitButtonClickType, onPlayButtonClickType);
export const privateRouteType = Object.assign({}, authoriationStatusType, exactType, pathType, renderType);
