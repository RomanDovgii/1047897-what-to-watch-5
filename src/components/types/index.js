import PropTypes from "prop-types";

export const appType = {
  name: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  releaseDate: PropTypes.number.isRequired,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  movies: PropTypes.arrayOf(PropTypes.string).isRequired
};

export const movieListItemType = PropTypes.string.isRequired;

export const movieListType = PropTypes.arrayOf(movieListItemType).isRequired;

export const movieCardTopType = {
  name: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  releaseDate: PropTypes.number.isRequired
};

export const genresListItemType = PropTypes.string.isRequired;

export const genresListType = PropTypes.arrayOf(genresListItemType).isRequired;

export const catalogType = {
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  movies: PropTypes.arrayOf(PropTypes.string).isRequired
};
