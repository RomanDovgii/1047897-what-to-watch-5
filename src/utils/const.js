export const MOVIES_COUNT = 9;
export const MAXIMUM_RAITING = 10;
export const MINIMUM_STARRING = 5;
export const MAXIMUM_STARRING = 10;
export const MAXIMUM_RAITING_COUNT = 1000;
export const MINIMUM_DESCRIPTION = 1;
export const MAXIMUM_DESCRIPTION = 10;

export const GENRES = [`All genres`, `Comedies`, `Crime`, `Documentary`, `Dramas`, `Horror`, `Kids & Family`, `Romance`, `Sci-Fi`, `Thrillers`];
export const MOVIES = [`Aviator`, `Bohemian Rhapsody`, `Dardjeeling Limited`, `Fantastic Beasts The Crimes of Grindelwald`, `Johnny English`, `Macbeth`, `Midnight Special`, `Mindhunter`, `Moonrise Kingdom`, `No Country For Old Men`, `Orlando`, `Pulp Fiction`, `Revenant`, `Seven Years In Tibet`, `Shutter Island`, `Snatch`, `War of the Worlds`, `We need to talk about Kevin`, `What We Do in the Shadows`];
export const VIDEO_URLS = [`https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`, `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`];
export const DIRECTORS = [`Steven Spielberg`, `Martin Scorsese`, `Alfred Hitchcock`, `Quentin Tarantino`, `Stanley Kubrick`, `Christopher Nolan`, `Tim Burton`, `Spike lee`];
export const ACTORS = [`Tom Hanks`, `Will Smith`, `Leonardo DiCaprio`, `Robert De Niro`, `Tom Cruise`, `Samuel L. Jackson`, `Robert Downey, Jr.`, `Scarlett Johansson`, `Jennifer Lawrence`, `Nickole Kidman`, `Cate Blanchett`, `Meryl Streep`, `Charlize Theron`, `Amy Adams`, `Angelina Jolie`];

export const FISH_TEXT = `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc sodales convallis elit et lacinia. Quisque quis cursus lectus. Praesent elementum semper felis ut pretium. Ut eu mi mollis, luctus purus ac, aliquet ipsum. Maecenas libero nunc, faucibus a porttitor eget, laoreet hendrerit diam. Aliquam eget purus sit amet augue feugiat efficitur in et augue. Curabitur nec nulla a mauris maximus semper vel in dui. Vivamus mollis urna ex, id facilisis velit facilisis tincidunt. Maecenas vitae congue massa. Maecenas vel tortor justo.`;

export const GRAND_BUDAPEST_HOTEL = {
  name: `The Grand Budapest Hotel`,
  genre: `Drama`,
  release: 2014,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Tom Cruise`],
  rating: 8.9,
  ratingsCount: 240,
  videoUrl: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  isMyList: false
};

export const CatalogHeadingVariant = {
  CATALOG: `Catalog`, // скрыт
  MOVIE_PAGE: `More like this` // не скрыт
};
