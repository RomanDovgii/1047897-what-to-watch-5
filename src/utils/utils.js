import React from "react";
import {GENRES, MOVIES, VIDEO_URLS, DIRECTORS, ACTORS, MOVIES_COUNT, MAXIMUM_RATING, MAXIMUM_RATING_COUNT, FISH_TEXT, ActorsCount, Description, Paragraph, Duration, Id, CommentCount, MAXIMUM_SIMIALAR_MOVIES, MAXIMUM_DISPLAYED_ACTORS, RatingSystem, ALL_GENRE} from "./const";

const getRandomElementFromArray = (array) => {
  const max = array.length;
  const randomNumber = Math.floor(Math.random() * max);

  return array[randomNumber];
};

const getRandomIntegerNumber = (firstNumber, secondNumber = 0) => {
  const max = Math.max(firstNumber, secondNumber);
  const min = Math.min(firstNumber, secondNumber);

  return Math.floor(Math.random() * (max - min) + min);
};

const getRandomBoolean = () => {
  return (Math.random() >= 0.5);
};

const generateRating = () => {
  return Math.round((Math.random() * MAXIMUM_RATING) * 1e1) / 1e1;
};

const generateActors = () => {
  const actors = [];

  const starringCount = getRandomIntegerNumber(ActorsCount.MINIMUM, ActorsCount.MAXIMUM);

  for (let i = 0; i < starringCount; i++) {
    actors.push(getRandomElementFromArray(ACTORS));
  }

  return actors;
};

const generateParagraphText = () => {
  const randomSentences = [];
  const sentences = FISH_TEXT.split(`. `);
  const descriptionLength = getRandomIntegerNumber(Description.MINIMUM, Description.MAXIMUM);

  for (let i = 0; i < descriptionLength; i++) {
    randomSentences.push(getRandomElementFromArray(sentences));
  }

  const text = randomSentences.join(`. `);

  return text;
};

const generateDescriptionParagraphs = () => {
  const paragraphsCount = getRandomIntegerNumber(Paragraph.MINIMUM, Paragraph.MAXIMUM);
  let paragraphs = [];

  for (let i = 0; i < paragraphsCount; i++) {
    paragraphs.push(generateParagraphText());
  }

  return paragraphs;
};

const generateRelease = () => {
  const date = new Date();

  return date.getFullYear();
};

const generateMovie = () => {
  return {
    name: getRandomElementFromArray(MOVIES),
    genre: getRandomElementFromArray(GENRES),
    release: generateRelease(),
    descriptionParagraphs: generateDescriptionParagraphs(),
    director: getRandomElementFromArray(DIRECTORS),
    actors: generateActors(),
    rating: generateRating(),
    ratingsCount: getRandomIntegerNumber(MAXIMUM_RATING_COUNT),
    videoUrl: getRandomElementFromArray(VIDEO_URLS),
    duration: getRandomIntegerNumber(Duration.MINIMUM, Duration.MAXIMUM),
    id: getRandomIntegerNumber(Id.MINIMUM, Id.MAXIMUM),
    isMyList: getRandomBoolean()
  };
};

export const generateComment = () => {
  return {
    text: generateParagraphText(),
    rating: generateRating(),
    name: getRandomElementFromArray(ACTORS),
    date: new Date()
  };
};

const generateComments = (count) => {
  return new Array(count).fill().map(generateComment);
};

export const generateCommentsById = (movieIds) => {
  const comments = movieIds.reduce((accumulator, id) => {
    const number = getRandomIntegerNumber(CommentCount.MINIMUM, CommentCount.MAXIMUM);
    accumulator[id] = generateComments(number);
    return accumulator;
  }, {});

  return comments;
};

export const filterMoviesForMyList = (movies) => {
  return movies.filter((movie) => movie.isMyList);
};

export const generateMovies = () => {
  return new Array(MOVIES_COUNT).fill().map(generateMovie);
};

export const generateVideoType = (url) => {
  const urlPieces = url.split(`.`);
  const lastIndex = urlPieces.length - 1;

  return `video/${urlPieces[lastIndex]}`;
};

export const filterMoviesByGenre = (movies, genre) => {
  const filteredMovies = movies.slice().filter((movie) => movie.genre === genre);

  let finalArray = filteredMovies;

  if (filteredMovies.length < MAXIMUM_SIMIALAR_MOVIES) {
    const difference = MAXIMUM_SIMIALAR_MOVIES - filterMoviesByGenre.length;
    const additionalMovies = movies.slice().filter((movie) => movie.genre !== genre).slice(difference);
    finalArray = filteredMovies.concat(additionalMovies);
  }

  return finalArray;
};

export const generateDurationString = (duration) => {
  return `${Math.floor(duration / 60)}h ${duration % 60}m`;
};

export const generateActorsText = (actors) => {
  const actorsMaxIndex = actors.length - 1;

  return (
    <span className="movie-card__details-value">
      {actors.map((actor, i) => (
        <React.Fragment key={i}>
          {actor}{i < actorsMaxIndex ? `, ` : ``} {i < actorsMaxIndex ? <br/> : ``}
        </React.Fragment>
      ))}
    </span>);
};

export const generateStarringString = (actors) => {
  let text = ``;
  let end = ``;
  let actorsLocal = actors.slice();

  if (actorsLocal.length > MAXIMUM_DISPLAYED_ACTORS) {
    actorsLocal = actors.slice(0, MAXIMUM_DISPLAYED_ACTORS);

    end = ` and other`;
  }

  text = actorsLocal.join(`, `);

  text = text.slice(0, -2);

  text += end;

  return text;
};

export const generateRatingText = (rating) => {
  switch (true) {
    case rating <= RatingSystem.BAD:
      return `Bad`;
    case rating <= RatingSystem.NORMAL:
      return `Normal`;
    case rating <= RatingSystem.GOOD:
      return `Good`;
    case rating < RatingSystem.AWESOME:
      return `Very good`;
    case rating === RatingSystem.AWESOME:
      return `Awesome`;
    default:
      return `Rating is incorrect`;
  }
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const createGenresList = (movies) => {
  const genresNotSorted = [];

  movies.map((movie) => {
    genresNotSorted.push(movie.genre);
  });

  const genres = [];

  genresNotSorted.map((genre) => {
    if (!genres.includes(genre)) {
      genres.push(genre);
    }
  });

  return [ALL_GENRE, ...genres];
};
