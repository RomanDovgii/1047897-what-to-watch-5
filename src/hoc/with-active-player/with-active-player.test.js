import React from "react";
import renderer from "react-test-renderer";
import {testChildrenType} from "../../components/types/types";
import {noop} from "../../utils/utils";
import withActivePlayer from "./with-active-player";

const testMovie = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `img/the-grand-budapest-hotel-poster.jpg`,
  previewImage: `img/the-grand-budapest-hotel.jpg`,
  backgroundImage: `img/the-grand-budapest-hotel-bg.jpg`,
  backgroundColor: `#ffffff`,
  videoLink: `https://some-link`,
  previewVideoLink: `https://some-link`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 8.9,
  scoresCount: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  runTime: 99,
  genre: `Comedy`,
  released: 2014,
  isFavorite: false
};

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = testChildrenType;

const MockComponentWrapped = withActivePlayer(MockComponent);

it(
    `Should render correct children`,
    () => {
      const tree = renderer
      .create(
          (<MockComponentWrapped
            movie = {testMovie}
            renderPlayer = {noop}
          >
            <React.Fragment/>
          </MockComponentWrapped>),
          {
            createNodeMock() {
              return {};
            }
          }
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
    }
);
