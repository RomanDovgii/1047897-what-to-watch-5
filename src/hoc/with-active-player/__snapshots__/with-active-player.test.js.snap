// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correct children 1`] = `<div />`;
